### PR TITLE
Add first/last to read specification

### PR DESCRIPTION
**Pull request by mpraglowski**
Now it will be possible:

```
first_event_in_dummy_stream = event_store.read.stream("Dummy").first
last_event_published = event_store.read.last
```

### CODE REVIEW

**Comment by mpraglowski** [comment]
@joelvh please review 
https://github.com/RailsEventStore/rails_event_store/pull/399/commits/d39bbd31d5dcc53c88e36d436b73070aac18f44b

**Comment by pawelpacana** [comment]
Getting closer including `Enumerable`.

File: ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb [hunk 1]
```diff
@@ -51,6 +51,18 @@ def read(direction, stream, from:, limit:, batch_size:)
                 .to_ary
             end
             BatchEnumerator.new(batch_size, limit || Float::INFINITY, reader).each
+          elsif read_as == RubyEventStore::Specification::FIRST
+            stream_entries
+              .ordered(direction, stream, offset_entry_id)
+              .combine(:event)
+              .map_with(:stream_entry_to_serialized_record, auto_struct: false)
+              .to_ary.first
```
Comment (joelvh) [inline ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb]: @mpraglowski I think you don't want `.to_ary` here -- just `first` should get you one record

File: ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb [hunk 2]
```diff
@@ -51,6 +51,18 @@ def read(direction, stream, from:, limit:, batch_size:)
...
+          elsif read_as == RubyEventStore::Specification::LAST
+            stream_entries
+              .ordered(direction, stream, offset_entry_id)
+              .combine(:event)
+              .map_with(:stream_entry_to_serialized_record, auto_struct: false)
+              .to_ary.last
```
Comment (joelvh) [inline ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb]: @mpraglowski I think you don't want `.to_ary` here -- just `last` should get you one record. Internally, it reverses the SQL `ORDER` and `LIMIT`s to one record. I'm not sure if that means of reversing the query is always accurate. Presumably your tests check for that.

File: ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb (hunk 1 above)
Comment (mpraglowski) [reply ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb]: right, `first` is implemented here

File: ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb (hunk 2 above)
Comment (mpraglowski) [reply ruby_event_store-rom/lib/ruby_event_store/rom/repositories/events.rb]: however here `last` without `to_ary` results in failing tests:

```
     NoMethodError:
       undefined method `last' for #<ROM::Relation::Composite:0x00007fd4226cecf8>
     Shared Example Group: :event_repository called from ./lib/ruby_event_store/spec/rom/event_repository_lint.rb:26
```

and

```
     Sequel::DatabaseError:
       SQLite3::SQLException: near "DESC": syntax error
     Shared Example Group: :event_repository called from ./lib/ruby_event_store/spec/rom/event_repository_lint.rb:26
```

**Comment by joelvh** [comment]
@mpraglowski worked up a fix to DRY things up. To solve the `last` issue, I switch direction and use `first`. Tests pass. However, can you make sure we have tests that sufficiently test first/last with an offset, to make sure the proper `last` event is retrieved?

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 3]
```diff
@@ -7,10 +7,16 @@ class Specification
     NO_LIMIT = Object.new.freeze
     # @private
     # @api private
-    NO_BATCH = Object.new.freeze
+    FIRST    = Object.new.freeze
```
Comment (paneq) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: just use `private_const` instead!

**Comment by joelvh** [comment]
thanks @mpraglowski !

File: rails_event_store-rspec/lib/rails_event_store/rspec/publish.rb [hunk 4]
```diff
@@ -79,7 +79,7 @@ def match_events?
       end
 
       def last_event(spec)
-        spec.backward.limit(1).each.first
+        spec.last
       end
```
Comment (pawelpacana) [inline rails_event_store-rspec/lib/rails_event_store/rspec/publish.rb]: inline method maybe?

File: rails_event_store_active_record-legacy/lib/rails_event_store_active_record/legacy/event_repository.rb [hunk 5]
```diff
@@ -76,7 +80,7 @@ def read(spec)
       private
 
       def total_limit(specification)
-        specification.limit? ? specification.count : Float::INFINITY
+        specification.limit
       end
```
Comment (pawelpacana) [inline rails_event_store_active_record-legacy/lib/rails_event_store_active_record/legacy/event_repository.rb]: would inline this as well

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 6]
```diff
@@ -149,10 +154,36 @@ def each
     # @param batch_size [Integer] number of events to read in a single batch
     # @return [Specification]
     def in_batches(batch_size = DEFAULT_BATCH_SIZE)
-      Specification.new(repository, mapper, result.dup.tap { |r| r.batch_size = batch_size })
+      Specification.new(repository, mapper, result.dup.tap { |r| r.read_as = :batch; r.batch_size = batch_size })
     end
     alias :in_batches_of :in_batches
 
+    def read_first
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: doc missing

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 7]
```diff
@@ -149,10 +154,36 @@ def each
...
+      Specification.new(repository, mapper, result.dup.tap { |r| r.read_as = :first })
+    end
+
+    def read_last
```
Comment (pawelpacana) [inline ruby_event_store/lib/ruby_event_store/specification.rb]: doc missing

File: ruby_event_store/spec/specification_spec.rb [hunk 8]
```diff
@@ -295,6 +356,38 @@ module RubyEventStore
...
+      expect(specification.from(records[2].event_id).first).to eq(TestEvent.new(event_id: records[3].event_id))
+      expect(specification.from(records[2].event_id).last).to eq(TestEvent.new(event_id: records[4].event_id))
+      expect(specification.from(records[2].event_id).backward.first).to eq(TestEvent.new(event_id: records[1].event_id))
+      expect(specification.from(records[2].event_id).backward.last).to eq(TestEvent.new(event_id: records[0].event_id))
+    end
+
+    specify do
+      repository.append_to_stream([test_record], Stream.new("Dummy"), ExpectedVersion.none)
+
+      expect(specification.result.batched?).to be_falsey
```
Comment (pawelpacana) [inline ruby_event_store/spec/specification_spec.rb]: eq(false)

File: ruby_event_store/spec/specification_spec.rb [hunk 9]
```diff
@@ -295,6 +356,38 @@ module RubyEventStore
...
+      expect(specification.result.first?).to be_falsey
+      expect(specification.result.last?).to be_falsey
+
+      expect(specification.read_first.result.batched?).to be_falsey
+      expect(specification.read_first.result.first?).to be_truthy
```
Comment (pawelpacana) [inline ruby_event_store/spec/specification_spec.rb]: eq(true)

**Pull request merged**